src/control/server/storage/scm/ipmctl_region.go

```diff
@@ -0,0 +1,376 @@
+	return nil
+}
+
+// checkIpmctl verifies ipmctl application version is acceptable.
+func (cr *cmdRunner) checkIpmctl(badList []semVer) (errOut error) {
+	cr.checkOnce.Do(func() {
+		cmdOut, err := cr.runCmd(cmdShowIpmctlVersion)
+		if err != nil {
+			errOut = errors.WithMessage(err, "show version cmd")
+			return
+		}
+
+		re := regexp.MustCompile(`(\d{2}).(\d{2}).(\d{2}).(\d{4})`)
+		matched := re.FindStringSubmatch(cmdOut)
+
+		if matched == nil {
+			errOut = errors.Errorf("could not read ipmctl version (%s)", cmdOut)
+			return
+		}
+
+		ipmctlBinVer := matched[1:]
+		cr.log.Debugf("ipmctl binary semver: %v", ipmctlBinVer)
+
+		errOut = validateSemVer(ipmctlBinVer, badList)
+	})
+
+	return
+}
+
+func (cr *cmdRunner) showRegions(sockID int) (string, error) {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return "", errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cmd := cmdShowRegions
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return "", errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return out, nil
+}
+
+func (cr *cmdRunner) createRegions(sockID int) error {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cr.log.Debug("set interleaved appdirect goal to create regions")
+
+	cmd := cmdCreateRegions
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return nil
+}
+
+func (cr *cmdRunner) removeRegions(sockID int) error {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cr.log.Debug("set memory mode goal to remove regions")
+
+	cmd := cmdRemoveRegions
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return nil
+}
+
+func (cr *cmdRunner) deleteGoals(sockID int) error {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cr.log.Debug("delete any existing memory allocation goals")
+
+	cmd := cmdDeleteGoals
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return nil
+}
+
+// socketRegionMap maps regions based on socket ID key.
+type socketRegionMap map[int]Region
+
+func (srm *socketRegionMap) keys() []int {
+	if srm == nil || len(*srm) == 0 {
+		return []int{}
+	}
+
+	keys := make([]int, 0, len(*srm))
+	for k := range *srm {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+
+	return keys
+}
```

Comment:
This will be changed to a non-pointer receiver, right?

Reply:
done